src/log/dmlog_appender.cpp

```diff
@@ -0,0 +1,65 @@
+#include <fc/log/dmlog_appender.hpp>
+#include <fc/log/log_message.hpp>
+#include <fc/string.hpp>
+#include <fc/variant.hpp>
+#include <fc/reflect/variant.hpp>
+#ifndef WIN32
+#include <unistd.h>
+#endif
+#include <boost/asio/io_context.hpp>
+#include <boost/thread/mutex.hpp>
+#include <fc/exception/exception.hpp>
+#include <iomanip>
+#include <mutex>
+#include <sstream>
+
+namespace fc {
+   class dmlog_appender::impl {
+      public:
+         boost::asio::io_service* io_service;
+         boost::mutex log_mutex;
+   };
+
+   dmlog_appender::dmlog_appender( const variant& args )
+   :my(new impl){}
+
+   dmlog_appender::dmlog_appender()
+   :my(new impl){}
+
+   dmlog_appender::~dmlog_appender() {}
+
+   void dmlog_appender::initialize( boost::asio::io_service& io_service ) {
+      my->io_service = &io_service;
+   }
+
+   void dmlog_appender::log( const log_message& m ) {
+      FILE* out = stdout;
+
+      string message = format_string( "DMLOG " + m.get_format() + "\n", m.get_data() );
+      std::unique_lock<boost::mutex> lock(my->log_mutex);
+
+      int retries = 0;
+      auto remaining_size = message.size();
+      auto message_ptr = message.c_str();
+      while (true) {
```

Comment:
For some reason, fwrite return zero (due to certain error), then retries causing asio stop, never can get out of the loop. 
if (written == remaining_size) alone is not good enough, please handle the error cases

Reply:
Indeed, good catch, the loop had `exit(1)` before which was getting out of the loop.

I changed this. After calling `stop`, I `return` immediately and set an `is_stopped` variable to `true`.

Then, upon the next `log` call, if stopped, I log a message indicating termination of the dmlog appender and return right away.